libgnucash/engine/gnc-rational-rounding.hpp

```diff
@@ -25,6 +25,8 @@
 #include "gnc-numeric.h"
 #include "gnc-int128.hpp"
 
+#define SIGN_NUM(number) ((number) == 0 ? 0 : (number) / std::abs(number))
```

Comment:
No macros please, this is C++. Use a template.
Did you profile ```number/std::abs(number)``` against ```number <  0 ? -1 :number > 0 ? 1 : 0```?

Reply:
Sorry, I am not that much comfortable with C++, but I guess it's alright now.

Reply:
```
template<typename T>
constexpr bool sign_number(T number)
{
    static_assert(std::is_arithmetic<T>::value);
    return number == std::abs(number);    
}
```

Reply:
@bhardwajs I ended up writing a template function that checks if the quotient between the two numbers is positive.

Reply:
@bhardwajs Not quite the same. ```SIGN_NUM``` returns -1, 0, or 1 and the usage is ```SIGN_NUM(numerator) * SIGN_NUM(denominator)```. That obviously won't work for a function that returns type bool. Also it's operating on runtime values so it can't be ```constexpr```.

Reply:
@DavPalma Looks good.

Reply:
Yes, @jralls , the usage `SIGN_NUM(numerator) * SIGN_NUM(denominator)` would require a numeric return type. Sorry about that!

For reference, `constexpr` can be used for functions operating on runtime values. For example, following code is valid and tested with `g++ -Wall -std=c++11`

```
template<typename T>
constexpr bool sign_number(T number)
{
    static_assert(std::is_arithmetic<T>::value);
    return number == std::abs(number);    
}

int main()
{
    int b;
    std::cin >> b;
    std::cout << sign_number(b) << '\n';
    
    return 0;
}
```

Reply:
@bhardwajs Well, sort of. If you don't use the function in a ```constexper``` context then the keyword is ignored. If instead you do
```
#include <iostream>
#include <type_traits>
#include <cmath>

template<typename T>
constexpr bool sign_number(T number)
{
    static_assert(std::is_arithmetic<T>::value);
    return number == std::abs(number);    
}

template<bool v>
struct constN
{
    constN() { std::cout << v << '\n'; }
};

int main()
{
    constN<sign_number(-17)> out;
    return 0;
}
```
So that it gets used as an actual constexpr and with Apple clang 10.0.1 you get
```
/Users/john/testconstexper.cpp:23:12: error: non-type template argument is not a
      constant expression
    constN<sign_number(-17)> out;
           ^~~~~~~~~~~~~~~~
/Users/john/testconstexper.cpp:9:22: note: non-constexpr function 'abs' cannot
      be used in a constant expression
    return number == std::abs(number);
                     ^
/Users/john/testconstexper.cpp:23:12: note: in call to 'sign_number(-17)'
    constN<sign_number(-17)> out;
```
Bottom line: Don't stick ```constexpr``` on template functions unless you're actually going to run the function at compile time, you might be introducing latent bugs.